src/proto/socks5address.cpp

```diff
@@ -22,33 +22,40 @@
 using namespace std;
 using namespace boost::asio::ip;
 
-int SOCKS5Address::parse(const string &data) {
+bool SOCKS5Address::parse(const string &data, size_t &address_len) {
     if (data.length() == 0 || (data[0] != IPv4 && data[0] != DOMAINNAME && data[0] != IPv6)) {
-        return -1;
+        return false;
     }
     address_type = static_cast<AddressType>(data[0]);
     switch (address_type) {
         case IPv4: {
-            if (data.length() >= 7) {
+            if (data.length() > 4 + 2) {
                 address = to_string(uint8_t(data[1])) + '.' +
                           to_string(uint8_t(data[2])) + '.' +
                           to_string(uint8_t(data[3])) + '.' +
                           to_string(uint8_t(data[4]));
                 port = (uint8_t(data[5]) << 8) | uint8_t(data[6]);
-                return 7;
+                address_len = 1 + 4 + 2;
+                return true;
             }
             break;
         }
         case DOMAINNAME: {
-            if (data.length() >= (unsigned int)(unsigned char)data[1] + 4) {
-                address = data.substr(2, data[1]);
-                port = (uint8_t(data[data[1] + 2]) << 8) | uint8_t(data[data[1] + 3]);
-                return data[1] + 4;
+            uint8_t domain_len = data[1];
+            if (domain_len < 1 || domain_len > 255) {
```

Comment:
It's impossible for an 8-bit number to exceed 255.
```suggestion
            if (domain_len == 0) {
```